docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,126 @@
+- new genesis parameters in `staking` module
+    - `MaxConsPubKeyRotations` : maximum number of rotation can be executed by a validator in recent unbonding period. default value 10 is suggested(11th key rotation will be rejected)
+    - `InitialKeyRotationFee` : the initial key rotation fee when no key rotation has happened in recent unbonding period. default value 1atom is suggested(1atom fee for the first key rotation in recent unbonding period)
+
+
+### Workflow
+
+1. The validator generates a new consensus keypair.
+2. The validator generates and signs a `MsgRotateConsPubKey` tx with their operator key and new ConsPubKey
+
+    ```go
+    type MsgRotateConsPubKey struct {
+        ValidatorAddress  sdk.ValAddress
+        NewPubKey         crypto.PubKey
+    }
+    ```
+
+3. `handleMsgRotateConsPubKey` gets `MsgRotateConsPubKey`, calls `RotateConsPubKey` with emits event
+4. `RotateConsPubKey` 
+    - checks if `NewPubKey` is not duplicated on `ValidatorsByConsAddr`
+    - checks if the validator is does not exceed parameter `MaxConsPubKeyRotations` by iterating `ConsPubKeyRotationHistory`
+    - checks if the signing account has enough balance to pay `KeyRotationFee`
+    - pays `KeyRotationFee` to community fund
+    - overwrites `NewPubKey` in `validator.ConsPubKey`
+    - deletes old `ValidatorByConsAddr`
+    - `SetValidatorByConsAddr` for `NewPubKey`
+    - Add `ConsPubKeyRotationHistory` for tracking rotation
+
+    ```go
+    type ConsPubKeyRotationHistory struct {
+        OperatorAddress         sdk.ValAddress
+        OldConsPubKey           crypto.PubKey
+        NewConsPubKey           crypto.PubKey
+        RotatedHeight           int64
+    }
+    ```
+
+5. `ApplyAndReturnValidatorSetUpdates` checks if there is `ConsPubKeyRotationHistory` with `ConsPubKeyRotationHistory.RotatedHeight == ctx.BlockHeight()` and if so, generates 2 `ValidatorUpdate` , one for a remove validator and one for create new validator 
+
+    ```go
+    abci.ValidatorUpdate{
+        PubKey: tmtypes.TM2PB.PubKey(OldConsPubKey),
+        Power:  0,
+    }
+
+    abci.ValidatorUpdate{
+        PubKey: tmtypes.TM2PB.PubKey(NewConsPubKey),
+        Power:  v.ConsensusPower(),
+    }
+    ```
+
+6. at `previousVotes` Iteration logic of `AllocateTokens`,  `previousVote` using `OldConsPubKey` match up with `ConsPubKeyRotationHistory`, and replace validator for token allocation
+7. Migrate `ValidatorSigningInfo` and `ValidatorMissedBlockBitArray` from `OldConsPubKey` to `NewConsPubKey`
+- Note : All above features shall be implemented in `staking` module.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- Validators can immediately or periodically rotate their consensus key to have better security policy
+- improved security against Long-Range attacks (https://nearprotocol.com/blog/long-range-attacks-and-a-new-fork-choice-rule) given a validator throws away the old consensus key(s)
+
+### Negative
```

Comment:
Missing an important negative of it allows a validator to effectively sell their entity, in a way that before would require hardware assumptions. (They can change their key to give it to new management, without any oversight from their delegators)